### PR TITLE
[SPARK-45136][CONNECT] Enhance ClosureCleaner with Ammonite support

### PR DESCRIPTION
**Pull request by vsevolodstep-db**
### What changes were proposed in this pull request?
This PR enhances existing ClosureCleaner implementation to support cleaning closures defined in Ammonite. Please refer to [this gist](https://gist.github.com/vsevolodstep-db/b8e4d676745d6e2d047ecac291e5254c) to get more context on how Ammonite code wrapping works and what problems I'm trying to solve here.

Overall, it contains these logical changes in `ClosureCleaner`:
1. Making it recognize and clean closures defined in Ammonite (previously it was checking if capturing class name starts with `$line` and ends with `$iw`, which is native Scala REPL specific thing
2. Making it clean closures if they are defined inside a user class in a REPL (see corner case 1 in the gist)
3. Making it clean nested closures properly for Ammonite REPL (see corner case 2 in the gist)
4. Making it transitively follow other Ammonite commands that are captured by the target closure.

Please note that `cleanTransitively` option of `ClosureCleaner.clean()` method refers to following references transitively within enclosing command object, but it doesn't follow other command objects.

As we need `ClosureCleaner` to be available in Spark Connect, I also moved the implementation to `common-utils` module. This brings a new `xbean-asm9-shaded` which is fairly small. 

Also, this PR moves `checkSerializable` check from `ClosureCleaner` to `SparkClosureCleaner`, as it is specific to Spark core

The important changes affect `ClosureCleaner` only. They should not affect existing codepath for normal Scala closures / closures defined in a native Scala REPL and cover only closures defined in Ammonite.

Also,  this PR modifies SparkConnect's `UserDefinedFunction` to actually use `ClosureCleaner` and clean closures in SparkConnect

### Why are the changes needed?
To properly support closures defined in Ammonite, reduce UDF payload size and avoid possible `NonSerializable` exceptions. This includes:
- lambda capturing outer command object, leading in a circular dependency
- lambda capturing other command objects transitively, exploding payload size

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests.
New tests in `ReplE2ESuite` covering various scenarios using SparkConnect + Ammonite REPL to make sure closures are actually cleaned. 

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: common/utils/pom.xml [hunk 1]
```diff
@@ -39,6 +39,10 @@
       <groupId>org.apache.spark</groupId>
       <artifactId>spark-tags_${scala.binary.version}</artifactId>
     </dependency>
+    <dependency>
```
[the remark this answers is not in the document]
Comment (vsevolodstep-db) [reply common/utils/pom.xml:42]: It's just a single jar containing shaded dependencies. It weights about 250KB

[the remark this answers is not in the document]
Comment (hvanhovell) [reply common/utils/pom.xml:42]: nice

File: common/utils/src/main/scala/org/apache/spark/util/SparkStreamUtils.scala [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+           |Current position $finalPos do not equal to expected position $expectedPos
+           |after transferTo, please check your kernel version to see if it is 2.6.32,
+           |this is a kernel bug which will lead to unexpected behavior when using transferTo.
+           |You can set spark.file.transferTo = false to disable this NIO feature.
+         """.stripMargin)
+    }
+  }
+}
+
+object SparkStreamUtils extends SparkStreamUtils
```
Comment (hvanhovell) [inline common/utils/src/main/scala/org/apache/spark/util/SparkStreamUtils.scala:109]: `private[spark]`?

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 3]
```diff
@@ -93,11 +93,12 @@ private[spark] object CallSite {
  * Various utility methods used by Spark.
  */
 private[spark] object Utils
-  extends Logging
-  with SparkClassUtils
-  with SparkErrorUtils
-  with SparkFileUtils
-  with SparkSerDeUtils {
+    extends Logging
```
Comment (hvanhovell) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:96]: Is this the proper formatting?

**Review by rednaxelafx — COMMENTED** [review]
(I just realized that I had some pending comments that I thought I had already submitted. Most of them are stale now but let me submit them anyway for reference)

File: common/utils/src/main/scala/org/apache/spark/util/ClosureCleaner.scala [hunk 4]
```diff
@@ -678,18 +851,45 @@ private[spark] object IndylambdaScalaClosures extends Logging {
     // Depth-first search for inner closures and track the fields that were accessed in them.
     // Start from the lambda body's implementation method, follow method invocations
     val visited = Set.empty[MethodIdentifier[_]]
-    val stack = Stack[MethodIdentifier[_]](implMethodId)
+    val queue = Queue[MethodIdentifier[_]](implMethodId)
```
[the remark this answers is not in the document]
Comment (vsevolodstep-db) [reply common/utils/src/main/scala/org/apache/spark/util/ClosureCleaner.scala:838]: Yes, this is intentional. Left a comment.
To be able to get references to all visited ammonite command objects we need to make sure that "parent" (in reference tree)  object is processed before the child

**Review by hvanhovell — APPROVED** [review]
LGTM - two nits.

**Comment by hvanhovell** [comment]
@vsevolodstep-db can you update the PR?

**Comment by hvanhovell** [comment]
Merging to master

**Pull request closed without merging**